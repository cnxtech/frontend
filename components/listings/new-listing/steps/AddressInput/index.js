import React, { Component } from 'react'
import { Formik, Field } from 'formik'
import Container from 'components/listings/new-listing/shared/Container'
import { ADDRESS_IS_COVERED } from 'graphql/listings/queries'
import Row from '@emcasa/ui-dom/components/Row'
import Col from '@emcasa/ui-dom/components/Col'
import Text from '@emcasa/ui-dom/components/Text'
import Input from '@emcasa/ui-dom/components/Input'
import StaticMap from 'components/listings/new-listing/shared/StaticMap'
import NavButtons from 'components/listings/new-listing/shared/NavButtons'
import { getAddressInput } from 'lib/address'
import {isMobile} from 'lib/mobile'
import AddressAutoComplete from 'components/shared/AddressAutoComplete'
import MobileAddressButton from 'components/shared/MobileAddressButton'

const Header = Text.withComponent('h1')

class AddressInput extends Component {
  constructor(props) {
    super(props)
    this.nextStep = this.nextStep.bind(this)
    this.updateLocation = this.updateLocation.bind(this)
    this.checkAddressCoverage = this.checkAddressCoverage.bind(this)
    this.validateAddress = this.validateAddress.bind(this)
    this.updateStateFromProps = this.updateStateFromProps.bind(this)
    this.onClearInput = this.onClearInput.bind(this)
    this.openMobileAddressInput = this.openMobileAddressInput.bind(this)
  }

  state = {
    address: null,
    complement: null,
    addressData: null,
    loading: false,
    error: null
  }

  componentDidMount() {
    this.updateStateFromProps(this.props)
  }

  componentWillReceiveProps(props) {
    this.updateStateFromProps(props)
  }

  updateStateFromProps(props) {
    const { location } = props
    if (location) {
      this.setState({
        address: location.address,
        complement: location.complement,
        addressData: location.addressData,
      })
    }
  }

  updateLocation() {
    const { updateLocation } = this.props
    updateLocation({
      address: this.state.address,
      complement: this.state.complement,
      addressData: this.state.addressData
    })
  }

  onClearInput() {
    this.setState({
      address: null,
      addressData: null
    })
  }

  nextStep() {
    const { navigateTo } = this.props
    navigateTo('homeDetails')
  }

  async checkAddressCoverage() {
    this.setState({loading: true})

    const input = getAddressInput(this.state.addressData)
    try {
      const { data } = await apolloClient.query({
        query: ADDRESS_IS_COVERED,
        variables: {
          city: input.city,
          neighborhood: input.neighborhood,
          state: input.state
        }
      })

      this.setState({loading: false})
      this.updateLocation()
      const { addressIsCovered } = data
      if (addressIsCovered) {
        this.nextStep()
      } else {
        const { navigateTo } = this.props
        navigateTo('notifyCoverage')
      }

    } catch (e) {
      this.setState({
        loading: false,
        error: 'Ocorreu um erro ao validar o endereço. Por favor, tente novamente.'
      })
    }
  }

  validateAddress(value) {
    if (!value) {
      return 'É necessário preencher um endereço'
    }
  }

  openMobileAddressInput() {
    this.props.navigateTo('addressInputMobile')
  }

  render() {
    const { location } = this.props
    let address, complement
    if (location) {
      address = location.address
      complement = location.complement
    }
    return (
      <div ref={this.props.hostRef}>
        <Container>
          <Col width={[1, 1/2]}>
            <Formik
              initialValues={{
                address: address,
                complement: complement
              }}
              isInitialValid={() => {
                return !(this.validateAddress(address))
              }}
              render={({isValid, setFieldValue, errors}) => (
                <>
                  <Row justifyContent="center">
                    <Header fontSize="large">
                      {this.props.title}
                    </Header>
                  </Row>
                  <Col>
                    <StaticMap animated={true} addressData={this.state.addressData} />
                  </Col>
                  {isMobile() ?
                    <Col mb={4}>
                      <MobileAddressButton
                        onClick={this.openMobileAddressInput}
                        address={address}
                      />
                    </Col>
                  :
                    <Col mb={4}>
                      <Field
                        name="address"
                        validate={this.validateAddress}
                        render={() => (
                          <AddressAutoComplete
                            defaultValue={address}
                            onClearInput={this.onClearInput}
                            onSelectAddress={(addressFormatted, addressData) => {
                              setFieldValue('address', addressFormatted)
                              this.setState({
                                address: addressFormatted,
                                addressData: addressData
                              })
                            }}
                          />
                        )}
                      />
                    </Col>
                  }
                  <Col>
                    <Field
                      name="complement"
                      render={() => (
                        <Input
                          hideLabelView
                          hideErrorView
                          placeholder="Complemento"
                          defaultValue={complement}
                          onChange={(e) => {
                            const { value } = e.target
                            setFieldValue('complement', value)
                            this.setState({complement: value})
                          }}
                        />
                      )}
                    />
                  </Col>
                  <NavButtons
                    onSubmit={() => {
                      this.checkAddressCoverage()
                    }}
                    submitEnabled={isValid}
                    loading={this.state.loading}
                  />
                </>
              )}
            />
          </Col>
        </Container>
      </div>
    )
  }
}

export default AddressInput
