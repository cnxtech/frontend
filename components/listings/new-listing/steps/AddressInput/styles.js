import styled from 'styled-components'
import { themeGet } from 'styled-system'
import Button from '@emcasa/ui-dom/components/Button'

const MobileAddressButton = styled(Button)`
  font-size: 18px;
  text-align: left;
  padding: 0 10px 0 10px;
  color: ${({hasAddress}) => hasAddress ? themeGet('colors.dark') : '#8291a8'};
  letter-spacing: 2px;
`

export {
  MobileAddressButton
}
