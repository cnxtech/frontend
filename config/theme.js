import theme from '@emcasa/ui'
import {fas} from '@fortawesome/pro-solid-svg-icons'
import {fal} from '@fortawesome/pro-light-svg-icons'
import icons from './icons'

export default {
  ...theme,
  icons: {emcasa: icons, default: fas, light: fal}
}
