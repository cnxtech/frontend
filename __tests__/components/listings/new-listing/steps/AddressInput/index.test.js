import React from 'react'
import AddressInput from 'components/listings/new-listing/steps/AddressInput'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import theme from 'config/theme'
import { ThemeProvider } from 'styled-components'

describe('<AddressInput/>', () => {
  it('should render a clean page', () => {
    const tree = renderer.create(
      <ThemeProvider theme={theme}>
        <AddressInput />
      </ThemeProvider>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
