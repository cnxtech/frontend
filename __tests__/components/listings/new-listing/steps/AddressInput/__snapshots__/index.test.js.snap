// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddressInput/> should render a clean page 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 20px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  width: 100%;
}

.c3 {
  margin-bottom: 20px;
}

.c6 {
  width: 100%;
}

.c10 {
  width: 41.66666666666667%;
}

.c2 {
  -webkit-font-smoothing: antialiased;
  font-family: FaktSoftPro-Normal;
  color: #3c4858;
  font-size: 24px;
  font-weight: bold;
  text-align: center;
}

.c8 {
  margin-right: 10px;
}

.c7 {
  -webkit-font-smoothing: antialiased;
  display: block;
  box-sizing: border-box;
  width: 100%;
  outline: none;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  padding: 0px 10px 0px 10px;
  height: 60px;
  color: #3c4858;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
}

.c7:focus {
  border-radius: 4px;
  border: 1px solid;
  border-color: #1fb6ff;
  padding: 0px 10px 0px 10px;
  height: 60px;
}

.c7::-webkit-input-placeholder {
  color: #8492a6;
}

.c7::-moz-placeholder {
  color: #8492a6;
}

.c7:-ms-input-placeholder {
  color: #8492a6;
}

.c7::placeholder {
  color: #8492a6;
}

.c4 {
  border: 1px solid #e0e6ed;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 {
  outline: none;
  display: block;
  cursor: pointer;
  padding: 0 20px 0 20px;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  background-color: #ffffff;
  width: 100%;
  height: 60px;
  color: #3c4858;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c11 :hover {
  background-color: #ffffff;
}

.c12 {
  outline: none;
  display: block;
  cursor: default;
  padding: 0 20px 0 20px;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  background-color: #ffffff;
  width: 100%;
  height: 60px;
  color: #8492a6;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c12 :hover {
  background-color: #ffffff;
}

@media screen and (min-width:640px) {
  .c1 {
    width: 50%;
  }
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
      width={
        Array [
          1,
          0.5,
        ]
      }
    >
      <p
        className="c2"
        color="dark"
        fontSize="large"
        fontWeight="bold"
      >
        Qual o endereço do seu imóvel?
      </p>
      <div
        className=""
      />
      <div
        className="c3"
      >
        <div
          className="c4 c5"
        >
          <div
            className="c6"
            width={1}
          >
            <div
              className=""
            >
              <input
                autoComplete="off"
                className="c7"
                fontSize="default"
                height="tall"
                letterSpacing="default"
                name="street"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyDown={[Function]}
                placeholder="Endereço e número*"
                style={
                  Object {
                    "border": 0,
                  }
                }
                type="text"
                value={null}
              />
            </div>
          </div>
          <div
            className="c8"
          />
        </div>
      </div>
      <div
        className=""
      >
        <div
          className=""
        >
          <input
            className="c7"
            fontSize="default"
            height="tall"
            letterSpacing="default"
            onChange={[Function]}
            placeholder="Complemento"
          />
        </div>
      </div>
      <div
        className="c9"
      >
        <div
          className="c10"
          width={0.4166666666666667}
        >
          <button
            className="c11"
            fontSize="default"
            height="tall"
            onClick={[Function]}
          >
            Voltar
          </button>
        </div>
        <div
          className="c10"
          width={0.4166666666666667}
        >
          <button
            className="c12"
            disabled={true}
            fontSize="default"
            height="tall"
            onClick={[Function]}
          >
            Avançar
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
